drivers/wifi/esp_at/esp_socket.c

```diff
@@ -76,7 +76,10 @@ void esp_socket_unref(struct esp_socket *sock)
 		}
 	} while (!atomic_cas(&sock->refcount, ref, ref - 1));
 
-	k_sem_give(&sock->sem_free);
+	/* notifies free only on 1-to-0 transition */
+	if (ref == 1) {
+		k_sem_give(&sock->sem_free);
+	}
```

Comment:
I suggest using the same pattern like in other parts of the system (for example in `net_pkt.c`) so it is easier to understand what is going on
```
if (ret > 1) {
	return;
}

k_sem_give(&sock->sem_free);
```

Reply:
follow it